### PR TITLE
Allow eliding GATs in expression position

### LINKED ISSUE #92836: Lifetime inference fails when construct then returning an instance of a GAT
I tried this code:

```rust
#![feature(generic_associated_types)]

pub trait Trait  {
    type Assoc<'a> where Self: 'a;

    fn f(&self) -> Self::Assoc<'_>;
}

pub struct Struct {
    item: f32
}

pub struct GenericStruct<'a> {
    ref_item: &'a f32
}

impl Trait for Struct {
    type Assoc<'a> = GenericStruct<'a>;

    fn f(&self) -> Self::Assoc<'_> {
        // This doesn't work
        Self::Assoc {
            ref_item: &self.item
        }
        
        // This doesn't work
        // Self::Assoc<'_> {
        //     ref_item: &self.item
        // }
        
        // This works, but concrete name is not always readily available
        // GenericStruct {
        //     ref_item: &self.item
        // }
        
        // This works
        // type Fix<'a> = <Struct as Trait>::Assoc<'a>;
        // Fix {
        //     ref_item: &self.item
        // }
    }
}
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ad49e24a0e96620b942d3abd23d49549

I expected to see this happen: the lifetime is correctly inferred when using
```rust
Self::Assoc {
    ref_item: &self.item
}
```

Instead, this happened: I have to use one of the work arounds described in the linked code.

Swap from GAT lifetime to the trait parameterized by a lifetime it works:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=551e676af610e612a7627661d6c5ca05

### Meta

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (1409c015b 2022-01-11)
binary: rustc
commit-hash: 1409c015b44a4d4d38bef2250b2a37c17b8b7463
commit-date: 2022-01-11
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by compiler-errors**
Thoughts on whether this is worthwhile?

Fixes #92836

r? @jackh726

### CODE REVIEW

File: compiler/rustc_typeck/src/astconv/mod.rs [hunk 1]
```diff
@@ -482,7 +482,21 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                        .astconv
+                        .re_infer(Some(param), self.span)
+                        .unwrap_or_else(|| {
+                            debug!(?param, "unelided lifetime in signature");
+
+                            // This indicates an illegal lifetime in a non-assoc-trait position
+                            tcx.sess
+                                .delay_span_bug(self.span, "unelided lifetime in signature");
+
+                            // Supply some dummy value. We don't have an
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/astconv/mod.rs:495]: copy-pasta'ed this from elsewhere in astconv.

**Comment by jackh726** [comment]
Will this change the lifetime we pick for something like `<Self as Trait>::Assoc`, where the definition of Trait has a lifetime?

**Comment by compiler-errors** [comment]
seems like (at least in expr position) we already unelide `<T as Trait>` into `<T as Trait<'_>>` where `'_` is a region variable: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d7ad0075064708258efbbb34d2b88c3f

so I don't... think so?

**Comment by jackh726** [comment]
This makes sense; lifetime elision is allowed here for non-GATs

@bors r+

**Pull request merged**